### PR TITLE
Add Formulas linter and integrate with factory

### PR DESCRIPTION
**Pull request by saumyaj3**
A new linter class, Formulas, has been added to check for issues in definition files, particularly with default parameters overrides. It has been integrated into the Linter factory to also check '.inst.cfg' and '.def.json' files for formulas-related issues. Additionally, a new 'diagnostic-incorrect-formula' check has been included in the .printer-linter configuration.

CURA-10901

### CODE REVIEW

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            file_data["overrides"]= overrides# Process the value here
+
+        return file_data
+
+    def checkValueIncorrect(self, formula) -> bool:
+        if isinstance(formula, str):
+            self._correct_formula = self._correctTyposInFormula(formula)
+            if self._correct_formula == formula:
+                return False
+            return True
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:139]: Minor optimization
```suggestion
            return self._correct_formula != formula
```

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    def _correctTyposInFormula(self, formula):
+        pattern = '|'.join(DELIMITERS)
+        tokens = re.split(pattern, formula)
+
+        output = formula
+        for token in tokens:
+            if '(' not in token and ')' not in token:
+                cleaned_token = re.sub(r'[^\w\s]', '', token)
+                possible_matches = difflib.get_close_matches(cleaned_token, self._cura_correction_strings, n=1, cutoff=0.8)
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:151]: :heart: :heart: :heart:

**Review by wawanbreton — COMMENTED** [review]
Looks good ! Some minor remarks, but the result is really cool.

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+FORMULA_NAMES = [
+    "extruderValue", "extruderValues", "anyExtruderWithMaterial", "anyExtruderNrWithOrDefault",
+    "resolveOrValue", "defaultExtruderPosition", "valueFromContainer", "extruderValueFromContainer"
+]
+
+DELIMITERS = [r'\+', '-', '=', '/', '\*', r'\(', r'\)', r'\[', r'\]', '{', '}', ' ', '^']
+
+
+class Formulas(Linter):
+    """Finds issues in definition files, such as overriding default parameters."""
```
[the remark this answers is not in the document]
Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py]: Yes. I guess I was sleepy :P  corrected here: fe0120ef64a924268095ccdffe319c5c59579df3

File: printer-linter/src/printerlinter/linters/formulas.py (hunk 1 above)
Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py]: Done here: fe0120ef64a924268095ccdffe319c5c59579df3

**Review by wawanbreton — APPROVED** [review]

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+FORMULA_NAMES = [
+    "extruderValue",
+    "extruderValues",
+    "anyExtruderWithMaterial",
+    "anyExtruderNrWithOrDefault",
+    "resolveOrValue",
+    "defaultExtruderPosition",
+    "valueFromContainer",
+    "extruderValueFromContainer",
+    "math",
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:22]: This is a module, not a function, right ?

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    "min",
+    "sqrt",
+    "log",
+    "tan",
+    "cos",
+    "sin",
+    "atan",
+    "acos",
+    "asin",
+    "pi",
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:35]: This is a constant so not callable as a function

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    "floor",
+    "debug",
+    "sum",
+    "len",
+    "uuid",
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:40]: This is also a module, not a function

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    "hashlib",
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:41]: This is also a module, not a function

File: printer-linter/src/printerlinter/linters/formulas.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    "base64",
+    "uuid3",
+    "NAMESPACE_DNS",
```
Comment (wawanbreton) [inline printer-linter/src/printerlinter/linters/formulas.py:44]: This is a constant so not callable as a function

**Review by wawanbreton — CHANGES_REQUESTED** [review]
Some added functions are not functions :)

File: printer-linter/src/printerlinter/linters/formulas.py (hunk 4 above)
Comment (wawanbreton) [reply printer-linter/src/printerlinter/linters/formulas.py:22]: Hmm so it's the same for the other `math` methods. Then maybe rename all the functions to add their module ? like `math.log`, `math.atan`, `math.acos`, ...

Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py:22]: I have '.' as a delimiter for strings.

File: printer-linter/src/printerlinter/linters/formulas.py (hunk 5 above)
Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py]: Removed all the non necessary constants, module or functions not used in printer definition here: https://github.com/Ultimaker/Cura/pull/18827/commits/72bf639beac129f6247a026d4c20cc56f2e2f372

File: printer-linter/src/printerlinter/linters/formulas.py (hunk 7 above)
Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py]: https://github.com/Ultimaker/Cura/pull/18827/commits/72bf639beac129f6247a026d4c20cc56f2e2f372

File: printer-linter/src/printerlinter/linters/formulas.py (hunk 8 above)
Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py]: https://github.com/Ultimaker/Cura/pull/18827/commits/72bf639beac129f6247a026d4c20cc56f2e2f372

File: printer-linter/src/printerlinter/linters/formulas.py (hunk 6 above)
Comment (saumyaj3) [reply printer-linter/src/printerlinter/linters/formulas.py]: https://github.com/Ultimaker/Cura/pull/18827/commits/72bf639beac129f6247a026d4c20cc56f2e2f372

**Review by wawanbreton — APPROVED** [review]

**Pull request merged**